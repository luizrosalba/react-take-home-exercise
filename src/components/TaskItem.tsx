import React from "react";
import { Task } from "./TaskManager";

type TaskItemProps = {
  task: Task;
  onDelete: (id: number)=> {};
  onToggle: (id: number)=> {}
}
const TaskItem = ({ task, onDelete, onToggle }: TaskItemProps) => {
  return (
    <li className="flex items-center justify-between border-b py-2">
      <span
        onClick={() => onToggle(task.id)}
        className={`cursor-pointer ${
          task.completed ? "line-through text-green-500" : "text-black"
        }`}
      >
        {task.title}
      </span>

      <button
        onClick={() => onDelete(task.id)}
        style={{
          backgroundColor: "red",
          color: "white",
          padding: "4px 8px",
          borderRadius: "4px",
        }}
      >
        Delete
      </button>
    </li>
  );
};

export default TaskItem;
