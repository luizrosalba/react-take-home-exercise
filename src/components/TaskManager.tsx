import React, { useState } from "react";

import TaskItem from "./TaskItem";
import { getInitialTasksFromLocalStorage, storeOnLocalStorage } from "../utils/localStorage";
import Button from "./Buttons/Button";

export type Task = { id: number, title: string, completed: boolean };
const TaskManager = () => {

  const [tasks, setTasks] = useState<Task[]>(getInitialTasksFromLocalStorage());
  const [filter, setFilter] = useState("all");
  const [newTask, setNewTask] = useState<string>("");

  // Intentional bug: The filter conditions are reversed.
  const filteredTasks = tasks.filter((task) => {
    if (filter === "completed") return task.completed === true;
    if (filter === "pending") return task.completed === false;
    return true;
  });

  const handleAddTask = (e: React.FormEvent) => {
    e.preventDefault();
    if (newTask) {
    if (newTask!.trim() === "") return;
      const newTaskObj = {
        id: tasks.length + 1,
        title: newTask,
        completed: false,
      };
      const updatedTasks = [...tasks, newTaskObj]
      storeOnLocalStorage({id: "storedTasks", value: updatedTasks})
      setTasks(updatedTasks);
      setNewTask("");
    }
  };

  // Intentional bug: Directly mutating the tasks array when deleting.
  const handleDeleteTask = (id: number) => {
      const task = tasks.find(task => task.id === id )
      if (task)
      {      
        const reponse = confirm(`Confirm to delete task: ${task?.title}`)
        if (reponse) {
          const updatedTasks = tasks.filter(task => task.id !== id )
          setTasks(updatedTasks);
          storeOnLocalStorage({id: "storedTasks", value: updatedTasks})
  
        }
      }
  };

  const toggleTaskCompletion = (id: number) => {
    const toggledTasks = tasks.map((task) => {
      if (task.id === id ) return {
        ...task, completed: !task.completed
      } 
      return task 
    });
    setTasks(toggledTasks)
  };

  return (
    <div className="container mx-auto bg-white p-4 rounded shadow">
      <form onSubmit={handleAddTask} className="mb-4 flex">
        <input
          type="text"
          placeholder="New task..."
          value={newTask}
          onChange={(e) => setNewTask(e.target.value)}
          className="flex-grow border rounded-l py-2 px-3"
        />
        <button type="submit" className="bg-blue-500 text-white px-4 rounded-r">
          Add
        </button>
      </form>
      <div className="flex justify-around mb-4">
        <Button 
          filter="all" 
          title={"All"} 
          setFilter={setFilter}
        /> 
        <Button 
          filter="completed" 
          title={"Completed"} 
          setFilter={setFilter}
        /> 
        <Button 
          filter="pending" 
          title={"Pending"} 
          setFilter={setFilter}
        /> 
      </div>
      <ul>
        {filteredTasks.map((task) => (
          <TaskItem
            key={task.id}
            task={task}
            onDelete={handleDeleteTask}
            onToggle={toggleTaskCompletion}
          />
        ))}
      </ul>
    </div>
  );
};

export default TaskManager;
