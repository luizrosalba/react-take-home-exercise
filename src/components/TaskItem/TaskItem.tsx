import { Task } from "../TaskManager/TaskManager";

type TaskItemProps = {
  task: Task;
  onDelete: (id: number) => void;
  onToggle: (id: number) => void
}
const TaskItem = ({ task, onDelete, onToggle }: TaskItemProps) => {
  return (
    <li className="flex items-center justify-between border-b py-2">
      <span
        data-testid="toggle-button"
        onClick={() => onToggle(task.id)}
        className={`cursor-pointer ${task.completed ? "line-through text-green-500" : "text-black"
          }`}
      >
        {task.title}
      </span>

      <button
        onClick={() => onDelete(task.id)}
        data-testid="delete-button"
        style={{
          backgroundColor: "red",
          color: "white",
          padding: "4px 8px",
          borderRadius: "4px",
        }}
      >
        Delete
      </button>
    </li>
  );
};

export default TaskItem;
