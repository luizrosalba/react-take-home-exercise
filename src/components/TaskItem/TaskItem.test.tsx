import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import TaskItem from "./TaskItem"
import { Task } from '../TaskManager/TaskManager';

describe('Filter Test', () => {
  const onDeleteMock = jest.fn()
  const onToggleMock = jest.fn()
  const completedTask = {
    completed: true,
    id: 1,
    title: 'Placeholder'
  } as Task
  it("Matches DOM Snapshot", () => {
    const { container } = render(<TaskItem onDelete={onDeleteMock} onToggle={onToggleMock} task={completedTask} key={'123'} />);
    expect(container).toMatchSnapshot();
  });
  it("Clicking Button should call mock", () => {
    render(<TaskItem onDelete={onDeleteMock} onToggle={onToggleMock} task={completedTask} key={'123'} />);
    const element = screen.queryByTestId('delete-button')
    element?.click();
    expect(onDeleteMock).toHaveBeenCalledTimes(1)
  });



});