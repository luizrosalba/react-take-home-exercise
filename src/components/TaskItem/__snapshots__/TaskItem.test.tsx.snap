// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Test Matches DOM Snapshot 1`] = `
<div>
  <li
    class="flex items-center justify-between border-b py-2"
  >
    <span
      class="cursor-pointer line-through text-green-500"
      data-testid="toggle-button"
    >
      Placeholder
    </span>
    <button
      data-testid="delete-button"
      style="background-color: red; color: white; padding: 4px 8px; border-radius: 4px;"
    >
      Delete
    </button>
  </li>
</div>
`;
